monkey/monkey_island/cc/repository/i_machine_repository.py

```diff
@@ -1,20 +1,74 @@
-        pass
-
-    # TODO define Machine object(ORM model)
-    # TODO define or re-use machine state.
-    # TODO investigate where should the state be stored in edge or both edge and machine
-    def get_machines(
-        self,
-        id: Optional[str] = None,
-        ips: Optional[Sequence[str]] = None,
-        state: Optional[MachineState] = None,  # noqa: F821
-        is_island: Optional[bool] = None,  # noqa: F841
-    ) -> Sequence[Machine]:  # noqa: F821
-        pass
+    """A repository used to store and retrieve Machines"""
+
+    @abstractmethod
+    def create_machine(self) -> Machine:
+        """
+        Create a new `Machine` in the repository
+
+        :return: A new `Machine` with a unique ID
+        :raises StorageError: If a new `Machine` could not be created
+        """
+
+    @abstractmethod
+    def update_machine(self, machine: Machine):
+        """
+        Update an existing `Machine` in the repository
+
+        :param machine: An updated Machine object to store in the repository
+        :raises UnknownRecordError: If the provided `Machine` does not exist in the repository
+        :raises StorageError: If an error occurred while attempting to store the `Machine`
+        """
+
+    @abstractmethod
+    def get_machine_by_id(self, machine_id: MachineID) -> Machine:
+        """
+        Get a `Machine` by ID
+
+        :param machine_id: The ID of the `Machine` to be retrieved
+        :return: A `Machine` with a matching `id`
+        :raises UnknownRecordError: If a `Machine` with the specified `id` does not exist in the
+                                    repository
+        :raises RetrievalError: If an error occurred while attempting to retrieve the `Machine`
+        """
+
+    @abstractmethod
+    def get_machine_by_hardware_id(self, hardware_id: HardwareID) -> Machine:
+        """
+        Get a `Machine` by ID
+
+        :param hardware_id: The hardware ID of the `Machine` to be retrieved
+        :return: A `Machine` with a matching `hardware_id`
+        :raises UnknownRecordError: If a `Machine` with the specified `hardware_id` does not exist
+                                    in the repository
+        :raises RetrievalError: If an error occurred while attempting to retrieve the `Machine`
+        """
+
+    @abstractmethod
+    def get_machines_by_ip(self, ip: IPv4Address) -> Sequence[Machine]:
+        """
+        Search for machines by IP address
+
+        :param ip: The IP address to search for
+        :return: A sequence of Machines that have a network interface with a matching IP
+        :raises UnknownRecordError: If a `Machine` with the specified `ip` does not exist in the
+                                    repository
+        :raises RetrievalError: If an error occurred while attempting to retrieve the `Machine`
+        """
+
+    @abstractmethod
+    def reset(self):
+        """
+        Removes all data from the repository
```

Comment:
It it deletes all data, then why not call it `remove_all_machines`?

Reply:
It would be nice if we had a protocol or interface, `Resetable` or some such, that would allow a component to accept a sequence of `Resetables` and reset them all. In less technical terms, it would be good if all repositories were as similar as possible. That way I don't need to remember or look up the name of the reset function for each repository every time I use it.

We could do `reset()`, `clear()`, or any number of other things. But let's try and pick something. We can go back through the other repositories and refactor them to comply.